File: Idil_addons/idil/static/src/js/pos_custom_product.js
odoo.define('Idil.PosCustomProducts', function (require) {
    'use strict';

    console.log("Custom POS Products JavaScript Loaded");  // Ensure this logs

    var models = require('point_of_sale.models');

    models.load_fields('product.product', ['name', 'list_price', 'qty_available', 'image_1920']);

    models.load_models([{
        model: 'my_product.product',
        fields: ['name', 'sale_price', 'stock_quantity', 'image_1920', 'available_in_pos'],
        domain: function(self) {
            return [['available_in_pos', '=', true]];
        },
        loaded: function(self, products) {
            console.log("Loaded custom products: ", products);  // Ensure this logs
            var product_db = self.db.add_products(_.map(products, function(product) {
                return {
                    id: product.id,
                    display_name: product.name,
                    list_price: product.sale_price,
                    qty_available: product.stock_quantity,
                    image_url: '/web/image?model=my_product.product&id=' + product.id + '&field=image_1920',
                };
            }));
            console.log("Custom Products added to POS DB: ", product_db);  // Ensure this logs
        },
    }], {'after': 'product.product'});
});

File: Idil_addons/idil/models/pos_custom_product.py
from odoo import models, fields, api


class PosCustomProduct(models.Model):
    _inherit = 'pos.session'

    @api.model
    def _loader_params_product_product(self):
        return {
            'search_params': {
                'fields': [
                    'id', 'name', 'display_name', 'sale_price', 'taxes_id',
                    'barcode', 'internal_reference', 'type', 'uom_id', 'sales_description',
                    'category_id', 'stock_quantity', 'available_in_pos'
                ],
                'domain': [('available_in_pos', '=', True)],
                'context': {'display_default_code': False},  # Ensure context is set
            },
        }

    def _get_products_from_ui(self, product_ids):
        params = self._loader_params_product_product()['search_params']
        context = params.get('context', {})
        fields = params['fields']
        products = self.env['my_product.product'].with_context(**context).search_read([('id', 'in', product_ids)],
                                                                                      fields)
        return products

    def _load_model(self, model_name):
        # Ensure context is always present
        params = self._loader_params_product_product()['search_params']
        context = params.get('context', {})
        self = self.with_context(**context)
        return super(PosCustomProduct, self)._load_model(model_name)
